init-tests-after-clone.sh

```diff
@@ -1,18 +1,73 @@
-#!/usr/bin/env bash
+#!/bin/sh
 
-set -e
+set -eu
 
-if [[ -z "$TRAVIS" ]]; then
-  read -rp "This operation will destroy locally modified files. Continue ? [N/y]: " answer
-  if [[ ! $answer =~ [yY] ]]; then
-    exit 2
-  fi
+fallback_repo_for_tags='https://github.com/gitpython-developers/GitPython.git'
+
+ci() {
+    # For now, check just these, as a false positive could lead to data loss.
+    test -n "${TRAVIS-}" || test -n "${GITHUB_ACTIONS-}"
+}
+
+no_version_tags() {
+    test -z "$(git tag -l '[0-9]*' 'v[0-9]*')"
+}
+
+warn() {
+    if test -n "${GITHUB_ACTIONS-}"; then
+        printf '::warning ::%s\n' "$*" >&2  # Annotate workflow.
+    else
+        printf '%s\n' "$@" >&2
+    fi
+}
+
+if ! ci; then
+    printf 'This operation will destroy locally modified files. Continue ? [N/y]: ' >&2
+    read -r answer
+    case "$answer" in
+    [yY])
+        ;;
+    *)
+        exit 2 ;;
+    esac
 fi
 
+# Stop if we have run this. (You can delete __testing_point__ to let it rerun.)
+# This also keeps track of where we are, so we can get back here.
 git tag __testing_point__
-git checkout master || git checkout -b master
+
+# The tests need a branch called master.
+git checkout master -- || git checkout -b master
+
+# The tests need a reflog history on the master branch.
 git reset --hard HEAD~1
 git reset --hard HEAD~1
 git reset --hard HEAD~1
+
+# Point the master branch where we started, so we test the correct code.
 git reset --hard __testing_point__
-git submodule update --init --recursive
+
+# The tests need submodules. (On CI, they would already have been checked out.)
+if ! ci; then
+    git submodule update --init --recursive
+fi
+
+# The tests need some version tags. Try to get them even in forks. This fetches
+# other objects too. So, locally, we always do it, for a consistent experience.
+if ! ci || no_version_tags; then
+    git fetch --all --tags
+fi
+
+# If we still have no version tags, try to get them from the original repo.
+if no_version_tags; then
+    warn 'No local or remote version tags found. Trying fallback remote:' \
+         "$fallback_repo_for_tags"
+
+    # git fetch supports * but not [], and --no-tags means no *other* tags, so...
+    printf 'refs/tags/%d*:refs/tags/%d*\n' 0 0 1 1 2 2 3 3 4 4 5 5 6 6 7 7 8 8 9 9 |
+        xargs git fetch --no-tags "$fallback_repo_for_tags"
```

Comment:
Going by [10.5 Git Internals - The Refspec](https://git-scm.com/book/en/v2/Git-Internals-The-Refspec), this `git fetch` command will fail prior to Git 2.6.0, which introduced the "partial glob" syntax where `*` need not be preceded immediately by a `/`.

I'm not sure if anything should be done about that. This is the last operation the script attempts, so little is lost due to its interaction with `set -e` if the feature is not supported. However, it should be possible to do this another way, using `git ls-remote` to list all the tags, filtering them, and then building a fetch command for the tags whose names match the regex `^v?[0-9]`. This would be more complicated to error check (`-o pipefail` is now POSIX, but [many `sh` implementations don't have it yet](https://github.com/koalaman/shellcheck/issues/2555)), but it could definitely be done. My bigger concern is that it would make this part of the script longer and more complicated.

It would also attempt to fetch a nonexistent tag in the rare case that a tag is deleted between when `ls-remote` run and when `git fetch` runs, but that's not very serious, and tags on this original repository, which the URL points to, are only rarely deleted. So this is the least import drawback.

(By the way, you may wonder why I have written such a belabored `printf` command. The reason is that [the `%n$` syntax](https://www.austingroupbugs.net/bug_view_page.php?bug_id=1592) for reusing an argument in the *same* format string is [not yet widely adopted](https://unix.stackexchange.com/a/744856/11938). Also, even if each number only had to be passed once, neither brace expansion, nor `seq`, nor `jot`, is guaranteed to be available.)

Reply:
I think this already is an improvement as is and am happy to keep it as long as there are no known issues. Those I wouldn't expect as, for whichever reason, the previous version also didn't see any issues that I would remember. Maybe people just don't report them, but I'd expect less people to not report them now.

Like mentioned before, I think it's absolutely possible, with less effort than it might initially seem, to take steps towards isolating tests and not require portions of this script anymore until it as a whole is not needed anymore. I am saying this merely from how it's done in `gitoxide`, and it's quite trivial once there is something like [`gix-testtools`](https://docs.rs/gix-testtools/0.12.0/gix_testtools/fn.scripted_fixture_read_only.html).